isis/src/base/apps/csminit/csminit.cpp

```diff
@@ -281,33 +291,63 @@ namespace Isis {
       kernelsGroup.deleteKeyword("Tolerance");
     }
 
-    // Remove tables from spiceinit
-    cube->deleteBlob("Table", "InstrumentPointing");
-    cube->deleteBlob("Table", "InstrumentPosition");
-    cube->deleteBlob("Table", "BodyRotation");
-    cube->deleteBlob("Table", "SunPosition");
-    cube->deleteBlob("Table", "CameraStatistics");
-    cube->deleteBlob("Polygon", "Footprint");
 
     if (cube->label()->hasObject("NaifKeywords")) {
       cube->label()->deleteObject("NaifKeywords");
     }
 
-    cube->deleteBlob("String", "CSMState");
-
     // Create our CSM State blob as a string
     // Add the CSM string to the Blob.
     StringBlob csmStateBlob(modelState, "CSMState");
     PvlObject &blobLabel = csmStateBlob.Label();
     blobLabel += PvlKeyword("ModelName", QString::fromStdString(model->getModelName()));
     blobLabel += PvlKeyword("PluginName", QString::fromStdString(plugin->getPluginName()));
 
-    // Write CSM State blob to cube
-    cube->write(csmStateBlob);
+    try {
+      CSMCamera(*cube, csmStateBlob);
+    } 
+    catch (IException &e) {
 
-    // TODO attempt to get the CSM Model from the cube
+      cube->deleteGroup("Instrument");
+      if (originalInstrument.keywords() != 0) {
+	cube->putGroup(originalInstrument);
+      }
 
+      cube->deleteGroup("Kernels");      
+      if (originalKernels.keywords() != 0) {
+	cube->putGroup(originalKernels);
+      }
+
+      cube->deleteGroup("CsmInfo");      
+      if (originalCsmInfo.keywords() != 0) {
+	cube->putGroup(originalCsmInfo);
+      }
+
+      QString message = "Failed to create a CSMCamera prior to writing the CSM state information to the cube.";
+      throw IException(e, IException::Unknown, message, _FILEINFO_);
+    }
+
+    // Delete existing CSMState Blob if it already exists before writing a new one
+    cube->deleteBlob("String", "CSMState");
+
+    // Remove tables from spiceinit before writing to the cube
+    cube->deleteBlob("Table", "InstrumentPointing");
+    cube->deleteBlob("Table", "InstrumentPosition");
+    cube->deleteBlob("Table", "BodyRotation");
+    cube->deleteBlob("Table", "SunPosition");
+    cube->deleteBlob("Table", "CameraStatistics");
+    cube->deleteBlob("Polygon", "Footprint");
+
+    // Write CSM State blob to cube
+    cube->write(csmStateBlob);
     p.WriteHistory(*cube);
-  }
 
+    try {
+      CameraFactory::Create(*cube);
```

Comment:
Should this call the CSMCamera Cube constructor to be consistent with the previous check?

Reply:
The first test is there to see if we can create the camera, this one is to make sure we wrote all the data to the cube file correctly and can create one from the file.